### PR TITLE
Support categories for events

### PR DESCRIPTION
**Pull request by ivarshav**
Support Categories for events (issue #29):

Add table
Add routers
Add tests
Frontend will be supported in issue #39 .

### CODE REVIEW

File: app/database/models.py [hunk 1]
```diff
@@ -64,12 +65,42 @@ class Event(Base):
...
+
+    @staticmethod
+    def create(db_session: Session, name: str, color: str, user_id: int):
+        try:
+            category = Category(name=name, color=color, user_id=user_id)
+            db_session.add(category)
+            db_session.flush()
+            db_session.commit()
+            db_session.refresh(category)
+            return category
```
Comment (yammesicka) [inline app/database/models.py:93]: This should be on the `else` clause

File: app/database/models.py [hunk 2]
```diff
@@ -64,12 +65,42 @@ class Event(Base):
...
+    __table_args__ = (
+        UniqueConstraint('user_id', 'name', 'color'),
+    )
+    id = Column(Integer, primary_key=True, index=True)
+    name = Column(String, nullable=False)
+    color = Column(String, nullable=False)
+    user_id = Column(Integer, ForeignKey("users.id"), nullable=False)
+
+    @staticmethod
+    def create(db_session: Session, name: str, color: str, user_id: int):
```
Comment (yammesicka) [inline app/database/models.py:86]: Please add the return type

File: app/routers/categories.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    else:
+        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST,
+                            detail=f"Request {request.query_params} contains "
+                                   f"unallowed params.")
+
+
+# TODO(issue#29): get current user_id from session
+@router.post("/")
+async def set_category(category: CategoryModel,
+                       db_session: Session = Depends(get_db)) -> Dict:
```
Comment (yammesicka) [inline app/routers/categories.py:40]: Please add more specification to the annotation regarding the values in the `Dict`

File: tests/test_categories.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    def test_get_categories_logic_succeeded(session, user, category):
+        assert get_user_categories(session, category.user_id) == [category]
+
+    @staticmethod
+    def test_creating_new_category(client, user):
+        response = client.post("/categories/",
+                               json={"user_id": user.id, "name": "Foo",
+                                     "color": "eecc11"})
+        assert response.status_code == status.HTTP_200_OK
+        assert {"user_id": user.id, "name": "Foo", "color": "eecc11"}.items() \
```
Comment (yammesicka) [inline tests/test_categories.py:25]: Prefer parenthesis :)

**Review by yammesicka — CHANGES_REQUESTED** [review]
Great job there. Added few insights, please take a look :)

File: tests/test_categories.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+    @staticmethod
+    def test_get_categories_logic_succeeded(session, user, category):
+        assert get_user_categories(session, category.user_id) == [category]
+
+    @staticmethod
+    def test_creating_new_category(client, user):
+        response = client.post("/categories/",
+                               json={"user_id": user.id, "name": "Foo",
+                                     "color": "eecc11"})
+        assert response.status_code == status.HTTP_200_OK
```
Comment (yammesicka) [inline tests/test_categories.py:24]: Prefer `.ok` to make this test pass even if the status code is 302 (or any other redirection)

File: app/database/models.py [hunk 6]
```diff
@@ -64,12 +69,44 @@ class Event(Base):
...
+    @staticmethod
+    def create(db_session: Session, name: str, color: str,
+               user_id: int) -> Category:
+        try:
+            category = Category(name=name, color=color, user_id=user_id)
+            db_session.add(category)
+            db_session.flush()
+            db_session.commit()
+            db_session.refresh(category)
+        except Exception as e:
```
Comment (yammesicka) [inline app/database/models.py:98]: Please catch a more specific exception here

File: tests/test_categories.py [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+    def test_get_categories_logic_succeeded(session, user, category):
+        assert get_user_categories(session, category.user_id) == [category]
+
+    @staticmethod
+    def test_creating_new_category(client, user):
+        response = client.post("/categories/",
+                               json={"user_id": user.id, "name": "Foo",
+                                     "color": "eecc11"})
+        assert response.ok
+        assert {"user_id": user.id, "name": "Foo",
```
Comment (yammesicka) [inline tests/test_categories.py:25]: I think it's better to leave out the `items()` call. The `json` might not be in the same order

**Review by yammesicka — CHANGES_REQUESTED** [review]
Awesome job. I have really few additional comments. Please take a look :)

File: tests/test_categories.py (hunk 7 above)
Comment (ivarshav) [reply tests/test_categories.py:25]: changed to set, so now order is not an issue ;)

**Comment by yammesicka** [comment]
Please fix the lint problems :)

**Comment by ivarshav** [comment]
> Please fix the lint problems :)

Thanks, done

**Pull request merged**